testing/rclone/APKBUILD

```diff
@@ -0,0 +1,98 @@
+# Contributor: tcely <rclone+aports@tcely.33mail.com>
+# Maintainer: tcely <rclone+aports@tcely.33mail.com>
+_pkgname="rclone"
+pkgname="${_pkgname}"
```

Comment:
This approach was used to address a concern from @Ikke about upstream pkgname not matching the sport name.

Reply:
Is upstream pkgname not "rclone"?

Reply:
It is, I understand that the concern is that URL and extracted directory name's not break when the sport is renamed.

Reply:
I don't understand - are there plans to rename it from the upstream name (which is the policy)?
How does this help in that scenario?
How does this not apply to literally every aport in the tree?

Reply:
I have no plans to rename it.
It helps by splitting aport name from upstream name and using the appropriate one.
This is the only aport in scope.

Reply:
So you're not going to rename it...
Upstream isn't going to rename it...
The two are currently the same...

What is so special about this aport that it needs this special treatment?

Reply:
I was made aware of two distinct usages of package name which just happen to be the same in most cases so I chose to use two variables. Please discuss this with @Ikke if you wish to know more.

Reply:
@tcely I think you missed the reason why I advised that way. The reason is you want to decouple the package name (which potentially can change) from the upstream name.

Say I want to make some alternative version of a package with some options or some modifications, I change the package name to make it unique, but now the package fails to build because it can not longer find the source.

This says to me that `pkgname` and the upstream name are two separate names, which at some point in time might happen to be the same.

By setting `_pkgname` to `$pkgname` and using that, you just couple those two again (though admittedly would only need to fix one variable).

Reply:
@Ikke I don't believe I missed the reason. I just don't see why I should assign twice from static text when the two variables are (currently) the same value. When renaming, just update the variable. Creating a single point of change for your case of renaming the aport is the whole reason for using two variables instead of just using `${pkgname}` as @5paceToast repeatedly suggested in the review.

Reply:
There are two possible states, with two distinct goals here.
Option 1: if the package is renamed, the APKBUILD should work as-is.
Option 2: the APKBUILD should be trivial when possible.

It is possible to choose Option 1, by having static assignments for both variations, allowing for a renaming of the package.
Under this Option, if someone was to change `pkgname` to `rclone-upstream`, the build would still work.
This, of course, is less trivial than Option 2.

It is also possible to choose Option 2, and minimize the complexity of the actual contents (what I was suggesting, since that is inherently more maintainable).
This means that one does not make a `_pkgname`, and would require changes in case of `pkgname` change.

Further, considering [your comment](https://github.com/alpinelinux/aports/pull/8248#discussion_r296954256), and the fact that Go applications are not trivially configurable, the odds of someone actually renaming the package are relatively slim, suggesting that Option 2 is the correct one.

However, your solution is neither.
It is non-trivial, and it does not allow the APKBUILD to function with a simple rename.
It is the worst of both worlds.

Reply:
> There are two possible states, with two distinct goals here.
> Option 1: if the package is renamed, the APKBUILD should work as-is.

This is more accurately stated as: if the package is renamed there should be no additional work beyond updating the single `pkgname` variable. This is silly in my opinion. This argument about having two distinct variables for the distinct usages of package name is also silly and a waste of time.

> Option 2: the APKBUILD should be trivial when possible.

You should probably define what you mean by trivial here.

> 
> It is possible to choose Option 1, by having static assignments for both variations, allowing for a renaming of the package.

Nothing about variable assignments allows or forbids package renaming, suggesting otherwise is a specious argument.

It is possible to assign the static text once and limit the work to the `pkgname` variable too.
```
_pkgname="rclone"
pkgname="${_pkgname}"
```

This is what I did when I first created the `APKBUILD` file.

> Under this Option, if someone was to change `pkgname` to `rclone-upstream`, the build would still work.
> This, of course, is less trivial than Option 2.

You will have to explain as to how you believe this is less trivial than option 2. It is not as apparent as you seem to believe.

> 
> It is also possible to choose Option 2, and minimize the complexity of the actual contents (what I was suggesting, since that is inherently more maintainable).
> This means that one does not make a `_pkgname`, and would require changes in case of `pkgname` change.
> 

This appears to suggest that maintenance is easier by foisting the decision of which is an upstream name and which is a downstream name onto the person coming along next. I don't agree with that notion at all.

> Further, considering [your comment](https://github.com/alpinelinux/aports/pull/8248#discussion_r296954256), and the fact that Go applications are not trivially configurable, the odds of someone actually renaming the package are relatively slim, suggesting that Option 2 is the correct one.
> 

I disagree with this too. Making the (unlikely) future change of the name (from either upstream or downstream) easier to address is the more correct choice in my opinion.

> However, your solution is neither.
> It is non-trivial, and it does not allow the APKBUILD to function with a simple rename.
> It is the worst of both worlds.

This is misstating the situation. I don't view changing one or two variables as non-trivial. Reviewing the entire `APKBUILD` for usages of `pkgname` and deciding on which name is appropriate is certainly non-trivial. Front loading that decision by using two variables is an entirely appropriate step to save a future contributor effort.

Reply:
> > There are two possible states, with two distinct goals here.
> > Option 1: if the package is renamed, the APKBUILD should work as-is.
> 
> This is more accurately stated as: if the package is renamed there should be no additional work beyond updating the single `pkgname` variable. This is silly in my opinion. This argument about having two distinct variables for the distinct usages of package name is also silly and a waste of time.

I agree!
So why are you doing it?
It's a waste of time and space, but you insist on continuing.

> > Option 2: the APKBUILD should be trivial when possible.
> 
> You should probably define what you mean by trivial here.

Trivial: Of little significance or value.
There is little less significant than nothingness.
Defining an extra variable is, by definition, less trivial than not defining it.
This is to minimize how many things one needs to have in mind when looking over the source.

> > It is possible to choose Option 1, by having static assignments for both variations, allowing for a renaming of the package.
> 
> Nothing about variable assignments allows or forbids package renaming, suggesting otherwise is a specious argument.

Of course, I don't mean that it is "impossible to rename the package".
I mean that it can be renamed without changing anything further...

> It is possible to assign the static text once and limit the work to the `pkgname` variable too.
> 
> ```
> _pkgname="rclone"
> pkgname="${_pkgname}"
> ```
> This is what I did when I first created the `APKBUILD` file.
> 
> > Under this Option, if someone was to change `pkgname` to `rclone-upstream`, the build would still work.
> > This, of course, is less trivial than Option 2.
> 
> You will have to explain as to how you believe this is less trivial than option 2. It is not as apparent as you seem to believe.

See above explanation of triviality.

> > It is also possible to choose Option 2, and minimize the complexity of the actual contents (what I was suggesting, since that is inherently more maintainable).
> > This means that one does not make a `_pkgname`, and would require changes in case of `pkgname` change.
> 
> This appears to suggest that maintenance is easier by foisting the decision of which is an upstream name and which is a downstream name onto the person coming along next. I don't agree with that notion at all.

There is no person coming along next.
I'm perfectly ok with actively maintaining my package, without someone else walking in - it's *why* I maintain them after all, to have recent versions available.
Further, there is no decision to make - there is the name of the folder inside the tarball, and there is whatever you call your package.
You don't decide which is which, they simply are.
Regarding maintenance, consider a person that gets confused when they see two identical variables, which are used in different scenarios.
They will try to figure out what's going on *every time they touch the package*, instead of maybe once in a very unlikely situation.

> > Further, considering [your comment](https://github.com/alpinelinux/aports/pull/8248#discussion_r296954256), and the fact that Go applications are not trivially configurable, the odds of someone actually renaming the package are relatively slim, suggesting that Option 2 is the correct one.
> 
> I disagree with this too. Making the (unlikely) future change of the name (from either upstream or downstream) easier to address is the more correct choice in my opinion.

You are allowed to have this opinion.
However, if I'm going to be the one handling the name of change (being the current maintainer), please do respect my desires regarding the package that I am going to be the one doing the work for.
See other comments regarding "up-to-dateness" regarding the obvious alternative.

> > However, your solution is neither.
> > It is non-trivial, and it does not allow the APKBUILD to function with a simple rename.
> > It is the worst of both worlds.
> 
> This is misstating the situation. I don't view changing one or two variables as non-trivial. Reviewing the entire `APKBUILD` for usages of `pkgname` and deciding on which name is appropriate is certainly non-trivial. Front loading that decision by using two variables is an entirely appropriate step to save a future contributor effort.

See above statement of triviality.
More trivial = less code.
More code = less trivial.
Therefore, it is less trivial than the alternative.
It not allowing for the process as described by @Ikke is also a matter of fact - one needs to change additional details (as it is now) in order to rename the package appropriately.

Reply:
> > > There are two possible states, with two distinct goals here.
> > > Option 1: if the package is renamed, the APKBUILD should work as-is.
> > 
> > 
> > This is more accurately stated as: if the package is renamed there should be no additional work beyond updating the single `pkgname` variable. This is silly in my opinion. This argument about having two distinct variables for the distinct usages of package name is also silly and a waste of time.
> 
> I agree!
> So why are you doing it?
> It's a waste of time and space, but you insist on continuing.
> 

I'm continuing because I don't agree with your assertions and do believe that they are worthy of a response.

> > > Option 2: the APKBUILD should be trivial when possible.
> > 
> > 
> > You should probably define what you mean by trivial here.
> 
> Trivial: Of little significance or value.
> There is little less significant than nothingness.
> Defining an extra variable is, by definition, less trivial than not defining it.
> This is to minimize how many things one needs to have in mind when looking over the source.
> 
> > > It is possible to choose Option 1, by having static assignments for both variations, allowing for a renaming of the package.
> > 
> > 
> > Nothing about variable assignments allows or forbids package renaming, suggesting otherwise is a specious argument.
> 
> Of course, I don't mean that it is "impossible to rename the package".
> I mean that it can be renamed without changing anything further...
> 

I already gave an example of how the two variable changes can be reduced to a single change and you gave another.

> > It is possible to assign the static text once and limit the work to the `pkgname` variable too.
> > ```
> > _pkgname="rclone"
> > pkgname="${_pkgname}"
> > ```
> > 
> > 
> > This is what I did when I first created the `APKBUILD` file.
> > > Under this Option, if someone was to change `pkgname` to `rclone-upstream`, the build would still work.
> > > This, of course, is less trivial than Option 2.
> > 
> > 
> > You will have to explain as to how you believe this is less trivial than option 2. It is not as apparent as you seem to believe.
> 
> See above explanation of triviality.
> 
> > > It is also possible to choose Option 2, and minimize the complexity of the actual contents (what I was suggesting, since that is inherently more maintainable).
> > > This means that one does not make a `_pkgname`, and would require changes in case of `pkgname` change.
> > 
> > 
> > This appears to suggest that maintenance is easier by foisting the decision of which is an upstream name and which is a downstream name onto the person coming along next. I don't agree with that notion at all.
> 
> There is no person coming along next.
> I'm perfectly ok with actively maintaining my package, without someone else walking in - it's _why_ I maintain them after all, to have recent versions available.
> Further, there is no decision to make - there is the name of the folder inside the tarball, and there is whatever you call your package.
> You don't decide which is which, they simply are.
> Regarding maintenance, consider a person that gets confused when they see two identical variables, which are used in different scenarios.
> They will try to figure out what's going on _every time they touch the package_, instead of maybe once in a very unlikely situation.
> 

I disagree with everything you've stated here.
You are not going to be the only person to ever touch any package you maintain.
The decision is which name should be used. You proved this yourself when you suggested an inappropriate `builddir` value earlier. Your pull request uses the GitHub generated archive which was named to match the default `builddir` value, where as the signed source archive contains a directory named differently.
These are not "two identical" variables. You could call them similar, but they are distinct variables. They currently do hold an identical value, so some confusion can be understood.

> > > Further, considering [your comment](https://github.com/alpinelinux/aports/pull/8248#discussion_r296954256), and the fact that Go applications are not trivially configurable, the odds of someone actually renaming the package are relatively slim, suggesting that Option 2 is the correct one.
> > 
> > 
> > I disagree with this too. Making the (unlikely) future change of the name (from either upstream or downstream) easier to address is the more correct choice in my opinion.
> 
> You are allowed to have this opinion.
> However, if I'm going to be the one handling the name of change (being the current maintainer), please do respect my desires regarding the package that I am going to be the one doing the work for.
> See other comments regarding "up-to-dateness" regarding the obvious alternative.
> 
> > > However, your solution is neither.
> > > It is non-trivial, and it does not allow the APKBUILD to function with a simple rename.
> > > It is the worst of both worlds.
> > 
> > 
> > This is misstating the situation. I don't view changing one or two variables as non-trivial. Reviewing the entire `APKBUILD` for usages of `pkgname` and deciding on which name is appropriate is certainly non-trivial. Front loading that decision by using two variables is an entirely appropriate step to save a future contributor effort.
> 
> See above statement of triviality.
> More trivial = less code.
> More code = less trivial.
> Therefore, it is less trivial than the alternative.
> It not allowing for the process as described by @Ikke is also a matter of fact - one needs to change additional details (as it is now) in order to rename the package appropriately.

It does allow for the changing of the package name (either upstream or downstream) as described, just not to your liking. The alternative is to use the same variable for every usage, which from a number of decisions and number of changes is not less trivial than one or two changed values in variables.

Reply:
> I'm continuing because I don't agree with your assertions and do believe that they are worthy of a response.

I can respect that.
I'm interested in figuring out the closest opinion to the truth™️.

> I already gave an example of how the two variable changes can be reduced to a single change and you gave another.

That's fair, so it ultimately fits the pre-emptive "what if mv happens" scenario.
However, in my opinion, "looking over the entirety of the APKBUILD" should be a requirement for modifying them in non-trivial ways (in this case, a "trivial" modification is a version bump and nothing else) - bad things can happen.
I would rather things within be simpler, rather than pre-emptively optimizing against unlikely scenarios.
However, the simple variant shouldn't make it impossible to later fix things (see: switch from libressl to openssl, proj4 renaming to proj).

> I disagree with everything you've stated here.
> You are not going to be the only person to ever touch any package you maintain.

My intention, when I start to maintain a package, is for no one to need to touch it except me.
I fully intend to update 100% of my maintained packages in a timely manner, and to support them in case bad things happen.
Sure, someone else could touch the packages I maintain, but my intention is to make that event wholly unnecessary.

> The decision is which name should be used. You proved this yourself when you suggested an inappropriate `builddir` value earlier. Your pull request uses the GitHub generated archive which was named to match the default `builddir` value, where as the signed source archive contains a directory named differently.
> These are not "two identical" variables. You could call them similar, but they are distinct variables. They currently do hold an identical value, so some confusion can be understood.

That's fair.
However, in the vast majority of cases, they do coincidentally match.
I don't think we need to pre-emptively optimize for an unlikely scenario at the expense of the contents (even if the expense is small), because one will accumulate over time (every time someone looks at/edits the build), and one is a simple one-time cost in an unlikely scenario.

> It does allow for the changing of the package name (either upstream or downstream) as described, just not to your liking. The alternative is to use the same variable for every usage, which from a number of decisions and number of changes is not less trivial than one or two changed values in variables.

The changes required to rename the packages are less trivial, I agree.
But if something should be optimized for, it's the general case, imo.